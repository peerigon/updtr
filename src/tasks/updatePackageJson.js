import createUpdatedPackageJson from "./util/createUpdatedPackageJson";
import Sequence from "./util/Sequence";

function lastChar(str) {
    return str.charAt(str.length - 1);
}

function stringify(newPackageJson, oldPackageJsonStr) {
    let newPackageJsonStr = JSON.stringify(newPackageJson, null, "  ");
    const lastCharFromOldPackageJson = lastChar(oldPackageJsonStr);

    // Preserve the new line character at the end if there was one
    if (lastCharFromOldPackageJson !== lastChar(newPackageJsonStr)) {
        newPackageJsonStr += lastCharFromOldPackageJson;
    }

    return newPackageJsonStr;
}

async function enhanceErrorMessage(fn, enhancedMessage) {
    try {
        return await fn();
    } catch (err) {
        err.message = enhancedMessage + err.message;
        throw err;
    }
}

export default (async function updatePackageJson(updtr, updateResults) {
    const sequence = new Sequence("update-package-json", updtr);
    let oldPackageJsonStr;

    sequence.start();

    const oldPackageJson = await enhanceErrorMessage(
        async () => {
            oldPackageJsonStr = await updtr.readFile("package.json");

            return JSON.parse(oldPackageJsonStr);
        },
        "Error while trying to read the package.json: "
    );

    const newPackageJson = createUpdatedPackageJson(
        oldPackageJson,
        updateResults,
        updtr.config
    );

    await enhanceErrorMessage(
        () =>
            updtr.writeFile(
                "package.json",
                stringify(newPackageJson, oldPackageJsonStr)
            ),
        "Error while trying to write the package.json: "
    );

    sequence.end({
        packageJson: newPackageJson,
    });
});
