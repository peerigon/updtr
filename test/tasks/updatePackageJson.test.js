import updatePackageJson from "../../src/tasks/updatePackageJson";
import FakeUpdtr from "../helpers/FakeUpdtr";
import pickEventNames from "../helpers/pickEventNames";
import {
    module1ToLatestSuccess,
    module2ToLatestFail,
} from "../fixtures/updateResults";
import { outdatedRegular } from "../fixtures/packageJsons";

describe("updatePackageJson()", () => {
    it("should read and write the package json", async () => {
        const updtr = new FakeUpdtr();
        const updateResults = [];

        updtr.readFile.resolves(JSON.stringify({}));
        updtr.writeFile.resolves();

        await updatePackageJson(updtr, updateResults);

        expect(updtr.readFile.calledWith("package.json")).toBe(true);
        expect(updtr.writeFile.calledWith("package.json")).toBe(true);
    });
    it("should save a package.json with expected shape to the cwd", async () => {
        const updtr = new FakeUpdtr();
        const updateResults = [module1ToLatestSuccess, module2ToLatestFail];

        updtr.readFile.resolves(
            JSON.stringify({
                dependencies: {
                    [module1ToLatestSuccess.name]: "1.0.0",
                    [module2ToLatestFail.name]: "1.0.x",
                },
                devDependencies: {
                    [module1ToLatestSuccess.name]: "~1.0.x",
                    [module2ToLatestFail.name]: "1.0.x",
                },
            })
        );
        updtr.writeFile.resolves();

        await updatePackageJson(updtr, updateResults);

        expect(updtr.readFile.calledWith("package.json")).toBe(true);
        expect(updtr.writeFile.calledWith("package.json")).toBe(true);

        const packageJson = JSON.parse(updtr.writeFile.getCall(0).args[1]);

        expect(packageJson).toMatchSnapshot();
    });
    it("should not alter the formatting", async () => {
        const updtr = new FakeUpdtr();
        const updateResults = [module1ToLatestSuccess];
        const oldPackageJson = outdatedRegular;

        updtr.readFile.resolves(oldPackageJson);
        updtr.writeFile.resolves();

        await updatePackageJson(updtr, updateResults);

        const newPackageJson = updtr.writeFile.getCall(0).args[1];

        expect(newPackageJson).toBe(
            oldPackageJson.replace(
                /\^1\.0\.0/,
                "^" + module1ToLatestSuccess.updateTo
            )
        );
    });
    it("should emit the expected events", async () => {
        const updtr = new FakeUpdtr();
        const updateResults = [module1ToLatestSuccess];
        const eventNames = [
            "update-package-json/start",
            "update-package-json/end",
        ];

        updtr.readFile.resolves(JSON.stringify({}));
        updtr.writeFile.resolves();

        await updatePackageJson(updtr, updateResults);

        expect(pickEventNames(eventNames, updtr.emit.args)).toEqual(eventNames);
    });
    describe("errors", () => {
        it("should enhance the error message in case the package json could not been read", async () => {
            const updtr = new FakeUpdtr();
            const updateResults = [];
            let givenErr;

            updtr.readFile.rejects(new Error("Oops"));

            try {
                await updatePackageJson(updtr, updateResults);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toBeInstanceOf(Error);
            expect(givenErr.message).toMatch(
                /Error while trying to read the package\.json: Oops/
            );
        });
        it("should enhance the error message in case the package json could not been parsed", async () => {
            const updtr = new FakeUpdtr();
            const updateResults = [];
            let givenErr;

            updtr.readFile.resolves("Invalid JSON");

            try {
                await updatePackageJson(updtr, updateResults);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toBeInstanceOf(SyntaxError);
            expect(givenErr.message).toMatch(
                /Error while trying to read the package\.json: Unexpected token I/
            );
        });
        it("should enhance the error message in case the package json could not been written", async () => {
            const updtr = new FakeUpdtr();
            const updateResults = [];
            let givenErr;

            updtr.readFile.resolves(JSON.stringify({}));
            updtr.writeFile.rejects(new Error("Oops"));

            try {
                await updatePackageJson(updtr, updateResults);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toBeInstanceOf(Error);
            expect(givenErr.message).toMatch(
                /Error while trying to write the package\.json: Oops/
            );
        });
    });
});
