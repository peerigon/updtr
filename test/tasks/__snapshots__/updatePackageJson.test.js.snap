// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updatePackageJson() should save a package.json with expected shape to the cwd 1`] = `
Object {
  "dependencies": Object {
    "updtr-test-module-1": "2.0.0",
    "updtr-test-module-2": "1.0.x",
  },
  "devDependencies": Object {
    "updtr-test-module-1": "~2.0.x",
    "updtr-test-module-2": "1.0.x",
  },
}
`;
